In [4]:
from wrf import ll_to_xy, getvar, extract_vars
from netCDF4 import Dataset

## `getvar`

You can use wrf-python's `getvar` function to pull a number of diagnostics (i.e. not included in the raw output) from the wrfout.

The list of available diagnostics can be found here:

https://wrf-python.readthedocs.io/en/latest/user_api/generated/wrf.getvar.html

### A quick example

In [6]:
wrf_path = '/data/shared_data/model/2024052012/wrfout_d03_2024-05-21_14:00:00'
wrfout = Dataset(wrf_path)
theta = getvar(wrfout, 'theta')
theta

<xarray.DataArray 'theta' (bottom_top: 50, south_north: 145, west_east: 320)> Size: 9MB
array([[[284.703  , 284.68655, 284.6701 , ..., 283.29666, 283.3515 ,
         283.40518],
        [284.71143, 284.69345, 284.67545, ..., 283.3216 , 283.37625,
         283.43018],
        [284.71982, 284.6999 , 284.68002, ..., 283.34464, 283.4006 ,
         283.45407],
        ...,
        [292.37326, 292.35535, 292.3196 , ..., 284.17508, 284.17404,
         284.1748 ],
        [292.4116 , 292.39517, 292.35425, ..., 284.17984, 284.1785 ,
         284.1791 ],
        [292.4351 , 292.41992, 292.39954, ..., 284.1841 , 284.183  ,
         284.18292]],

       [[286.0929 , 286.0647 , 286.0362 , ..., 284.30435, 284.36795,
         284.43564],
        [286.10233, 286.07245, 286.04147, ..., 284.32523, 284.39368,
         284.4643 ],
        [286.11023, 286.07788, 286.04456, ..., 284.3445 , 284.41974,
         284.49295],
...
        [499.02267, 498.94128, 498.86328, ..., 499.0091 , 498.9895 ,
         498.97534],
        [499.03952, 498.96277, 498.89197, ..., 499.08328, 499.0606 ,
         499.04144],
        [499.05536, 498.98895, 498.92474, ..., 499.15884, 499.13214,
         499.108  ]],

       [[508.47714, 508.4315 , 508.38895, ..., 506.63785, 506.6478 ,
         506.6572 ],
        [508.47165, 508.42197, 508.37863, ..., 506.63474, 506.64502,
         506.6545 ],
        [508.4673 , 508.416  , 508.37088, ..., 506.63083, 506.64203,
         506.65207],
        ...,
        [509.08392, 509.03668, 508.9956 , ..., 509.03226, 509.0327 ,
         509.03387],
        [509.0977 , 509.0522 , 509.0125 , ..., 509.08276, 509.08124,
         509.07947],
        [509.11096, 509.06995, 509.0312 , ..., 509.13257, 509.1289 ,
         509.12485]]], dtype=float32)
Coordinates:
    XLONG    (south_north, west_east) float32 186kB -3.986 -3.973 ... 0.01031
    XLAT     (south_north, west_east) float32 186kB 54.24 54.24 ... 55.28 55.28
    XTIME    float32 4B 1.56e+03
    Time     datetime64[ns] 8B 2024-05-21T14:00:00
Dimensions without coordinates: bottom_top, south_north, west_east
Attributes:
    FieldType:    104
    MemoryOrder:  XYZ
    description:  potential temperature
    units:        K
    stagger:      
    coordinates:  XLONG XLAT XTIME
    projection:   LambertConformal(stand_lon=-2.0, moad_cen_lat=57.0, truelat...

It's important to know what dimensions your data has when trying to interpret it, especially against observations. Here we can see that we have a three dimensional dataset, with 50 inthe vertical, 145 from south to north and 320 from west to east.

### Surface level variables

WRF's lowest model level is around 25m above ground level, so taking the data from here is NOT a valid representation of a surface measurement. `wrf-python` allows us to choose some variables at lower levels, for example *2m Temperature* 'T2' and *10m Wind Speed and Direction Rotated to Earth Coordinates* 'uvmet10_wspd_wdir'.

In [7]:
winds = getvar(wrfout, 'uvmet10_wspd_wdir')
winds

<xarray.DataArray 'uvmet10_wspd_wdir' (wspd_wdir: 2, south_north: 145,
                                       west_east: 320)> Size: 371kB
array([[[  2.2256646,   2.3035138,   2.3646095, ...,   3.5340753,
           3.3087008,   3.1506271],
        [  2.2524312,   2.3284426,   2.3960156, ...,   3.424167 ,
           3.2063549,   3.076988 ],
        [  2.2822225,   2.349981 ,   2.4175274, ...,   3.2833982,
           3.0980182,   3.013792 ],
        ...,
        [  2.7275124,   2.7928057,   3.0312154, ...,   5.37404  ,
           5.3660183,   5.36943  ],
        [  2.5702634,   2.6282427,   2.883499 , ...,   5.38269  ,
           5.3685675,   5.3681936],
        [  2.487719 ,   2.5277736,   2.6466677, ...,   5.3886695,
           5.372566 ,   5.3660207]],

       [[311.29892  , 311.7413   , 311.99298  , ...,  16.897846 ,
          22.888956 ,  29.1575   ],
        [312.38037  , 312.8896   , 313.06372  , ...,  18.810986 ,
          25.242262 ,  31.938633 ],
        [313.3729   , 313.79175  , 314.0627   , ...,  20.410236 ,
          27.745123 ,  34.839237 ],
        ...,
        [111.579185 , 111.96226  , 112.84004  , ...,  99.89101  ,
         100.11557  , 100.32318  ],
        [105.43783  , 106.66577  , 107.97914  , ...,  99.56931  ,
          99.76694  ,  99.97334  ],
        [105.03631  , 105.59359  , 106.52772  , ...,  99.1771   ,
          99.39223  ,  99.62783  ]]], dtype=float32)
Coordinates:
  * wspd_wdir  (wspd_wdir) <U4 32B 'wspd' 'wdir'
    XLONG      (south_north, west_east) float32 186kB -3.986 -3.973 ... 0.01031
    XLAT       (south_north, west_east) float32 186kB 54.24 54.24 ... 55.28
    XTIME      float32 4B 1.56e+03
    Time       datetime64[ns] 8B 2024-05-21T14:00:00
Dimensions without coordinates: south_north, west_east
Attributes:
    FieldType:    104
    MemoryOrder:  XY
    description:  10m earth rotated wspd,wdir
    units:        m s-1
    stagger:      
    coordinates:  XLONG XLAT XTIME
    projection:   LambertConformal(stand_lon=-2.0, moad_cen_lat=57.0, truelat...

What are our dimensions? The first is 2, this represents the two 2D arrays representing wind direction and wind speed. The following two dimensions are south to north and west to east, i.e. the horizontal WRF grid.

## Importing multiple files

Let's import WRF data for a whole day:

In [13]:
import os
import glob
path = '/data/shared_data/model/2024052012/'
os.chdir(path)
all_wrfouts = glob.glob('wrfout_d03*_2024-05-21*')
all_wrfouts.sort()
all_wrfouts

['wrfout_d03_2024-05-21_00:00:00',
 'wrfout_d03_2024-05-21_01:00:00',
 'wrfout_d03_2024-05-21_02:00:00',
 'wrfout_d03_2024-05-21_03:00:00',
 'wrfout_d03_2024-05-21_04:00:00',
 'wrfout_d03_2024-05-21_05:00:00',
 'wrfout_d03_2024-05-21_06:00:00',
 'wrfout_d03_2024-05-21_07:00:00',
 'wrfout_d03_2024-05-21_08:00:00',
 'wrfout_d03_2024-05-21_09:00:00',
 'wrfout_d03_2024-05-21_10:00:00',
 'wrfout_d03_2024-05-21_11:00:00',
 'wrfout_d03_2024-05-21_12:00:00',
 'wrfout_d03_2024-05-21_13:00:00',
 'wrfout_d03_2024-05-21_14:00:00',
 'wrfout_d03_2024-05-21_15:00:00',
 'wrfout_d03_2024-05-21_16:00:00',
 'wrfout_d03_2024-05-21_17:00:00',
 'wrfout_d03_2024-05-21_18:00:00',
 'wrfout_d03_2024-05-21_19:00:00',
 'wrfout_d03_2024-05-21_20:00:00',
 'wrfout_d03_2024-05-21_21:00:00',
 'wrfout_d03_2024-05-21_22:00:00',
 'wrfout_d03_2024-05-21_23:00:00']

We can extract a variables from all of these files into a single netcdf object with a new dimension: time.

In [18]:
datasets = [Dataset(file) for file in all_wrfouts]
T2 = getvar(datasets, 'T2', timeidx=None, method='cat')
T2

<xarray.DataArray 'T2' (Time: 24, south_north: 145, west_east: 320)> Size: 4MB
array([[[285.3145 , 285.31204, 285.3102 , ..., 283.72318, 283.7053 ,
         283.68585],
        [285.3352 , 285.3335 , 285.33197, ..., 283.72412, 283.707  ,
         283.68814],
        [285.35104, 285.3496 , 285.3488 , ..., 283.72418, 283.7082 ,
         283.68994],
        ...,
        [280.4543 , 280.45746, 280.31583, ..., 284.30777, 284.31528,
         284.322  ],
        [280.35037, 280.37112, 280.15283, ..., 284.31055, 284.3181 ,
         284.32452],
        [280.67465, 280.71808, 280.55234, ..., 284.31308, 284.32037,
         284.32703]],

       [[285.17712, 285.17654, 285.17615, ..., 283.7472 , 283.72864,
         283.70856],
        [285.1941 , 285.1951 , 285.1964 , ..., 283.74832, 283.7303 ,
         283.71146],
        [285.20718, 285.21014, 285.21375, ..., 283.7486 , 283.73172,
         283.71317],
...
        [284.178  , 284.21194, 284.11368, ..., 284.80185, 284.8123 ,
         284.82083],
        [284.10077, 284.20135, 284.01968, ..., 284.80734, 284.81705,
         284.82602],
        [284.34235, 284.31854, 284.17844, ..., 284.813  , 284.82257,
         284.83133]],

       [[285.38858, 285.3711 , 285.35208, ..., 284.0879 , 284.07242,
         284.0565 ],
        [285.38632, 285.36984, 285.35175, ..., 284.09348, 284.07812,
         284.06198],
        [285.37546, 285.3589 , 285.34152, ..., 284.09592, 284.0813 ,
         284.0656 ],
        ...,
        [283.56256, 283.62006, 283.55032, ..., 284.8217 , 284.83447,
         284.845  ],
        [283.50354, 283.50946, 283.35022, ..., 284.8288 , 284.84073,
         284.85123],
        [283.53537, 283.54834, 283.47574, ..., 284.83408, 284.84595,
         284.85712]]], dtype=float32)
Coordinates:
    XLONG     (south_north, west_east) float32 186kB -3.986 -3.973 ... 0.01031
    XLAT      (south_north, west_east) float32 186kB 54.24 54.24 ... 55.28 55.28
    XTIME     (Time) float64 192B 720.0 780.0 840.0 ... 2.04e+03 2.1e+03
  * Time      (Time) datetime64[ns] 192B 2024-05-21 ... 2024-05-21T23:00:00
    datetime  (Time) datetime64[ns] 192B 2024-05-21 ... 2024-05-21T23:00:00
Dimensions without coordinates: south_north, west_east
Attributes:
    FieldType:    104
    MemoryOrder:  XY 
    description:  TEMP at 2 M
    units:        K
    stagger:      
    coordinates:  XLONG XLAT XTIME
    projection:   LambertConformal(stand_lon=-2.0, moad_cen_lat=57.0, truelat...

Let's look at the new time dimension:

In [20]:
T2['Time']

<xarray.DataArray 'Time' (Time: 24)> Size: 192B
array(['2024-05-21T00:00:00.000000000', '2024-05-21T01:00:00.000000000',
       '2024-05-21T02:00:00.000000000', '2024-05-21T03:00:00.000000000',
       '2024-05-21T04:00:00.000000000', '2024-05-21T05:00:00.000000000',
       '2024-05-21T06:00:00.000000000', '2024-05-21T07:00:00.000000000',
       '2024-05-21T08:00:00.000000000', '2024-05-21T09:00:00.000000000',
       '2024-05-21T10:00:00.000000000', '2024-05-21T11:00:00.000000000',
       '2024-05-21T12:00:00.000000000', '2024-05-21T13:00:00.000000000',
       '2024-05-21T14:00:00.000000000', '2024-05-21T15:00:00.000000000',
       '2024-05-21T16:00:00.000000000', '2024-05-21T17:00:00.000000000',
       '2024-05-21T18:00:00.000000000', '2024-05-21T19:00:00.000000000',
       '2024-05-21T20:00:00.000000000', '2024-05-21T21:00:00.000000000',
       '2024-05-21T22:00:00.000000000', '2024-05-21T23:00:00.000000000'],
      dtype='datetime64[ns]')
Coordinates:
    XTIME     (Time) float64 192B 720.0 780.0 840.0 ... 2.04e+03 2.1e+03
  * Time      (Time) datetime64[ns] 192B 2024-05-21 ... 2024-05-21T23:00:00
    datetime  (Time) datetime64[ns] 192B 2024-05-21 ... 2024-05-21T23:00:00

## Extract vars

This function allows you to pull variables directly from the wrfout, i.e. variables that are produced natively by WRF rather than needing to be calculated as with the above examples. The input and output are slightly different, as you can provide multiple variables in a list which will return a python dictionary, like so:

In [27]:
COORDS = extract_vars(datasets, varnames=['XLAT', 'XLONG'], timeidx=None)

In [28]:
COORDS

{'XLAT': <xarray.DataArray 'XLAT' (Time: 24, south_north: 145, west_east: 320)> Size: 4MB
 array([[[54.241455, 54.241665, 54.241875, ..., 54.242336, 54.242134,
          54.24193 ],
         [54.248676, 54.24889 , 54.2491  , ..., 54.249554, 54.249355,
          54.249146],
         [54.255898, 54.256107, 54.256313, ..., 54.256783, 54.256573,
          54.25637 ],
         ...,
         [55.267124, 55.267334, 55.267555, ..., 55.268032, 55.26782 ,
          55.267605],
         [55.274345, 55.274563, 55.274776, ..., 55.275253, 55.27504 ,
          55.27483 ],
         [55.28157 , 55.281788, 55.282005, ..., 55.282482, 55.28227 ,
          55.28205 ]],
 
        [[54.241455, 54.241665, 54.241875, ..., 54.242336, 54.242134,
          54.24193 ],
         [54.248676, 54.24889 , 54.2491  , ..., 54.249554, 54.249355,
          54.249146],
         [54.255898, 54.256107, 54.256313, ..., 54.256783, 54.256573,
          54.25637 ],
 ...
         [55.267124, 55.267334, 55.267555, ..., 55.268032, 5

So we can access this data using the variable name as a key:

In [29]:
COORDS['XLONG']

<xarray.DataArray 'XLONG' (Time: 24, south_north: 145, west_east: 320)> Size: 4MB
array([[[-3.9858093e+00, -3.9734802e+00, -3.9611206e+00, ...,
         -6.6741943e-02, -5.4412842e-02, -4.2053223e-02],
        [-3.9861755e+00, -3.9738159e+00, -3.9614563e+00, ...,
         -6.6406250e-02, -5.4046631e-02, -4.1687012e-02],
        [-3.9865417e+00, -3.9741821e+00, -3.9618225e+00, ...,
         -6.6070557e-02, -5.3710938e-02, -4.1320801e-02],
        ...,
        [-4.0381775e+00, -4.0255127e+00, -4.0128174e+00, ...,
         -1.5777588e-02, -3.1127930e-03,  9.5825195e-03],
        [-4.0385742e+00, -4.0258789e+00, -4.0131836e+00, ...,
         -1.5411377e-02, -2.7160645e-03,  9.9487305e-03],
        [-4.0389404e+00, -4.0262451e+00, -4.0135803e+00, ...,
         -1.5045166e-02, -2.3498535e-03,  1.0314941e-02]],

       [[-3.9858093e+00, -3.9734802e+00, -3.9611206e+00, ...,
         -6.6741943e-02, -5.4412842e-02, -4.2053223e-02],
        [-3.9861755e+00, -3.9738159e+00, -3.9614563e+00, ...,
         -6.6406250e-02, -5.4046631e-02, -4.1687012e-02],
        [-3.9865417e+00, -3.9741821e+00, -3.9618225e+00, ...,
         -6.6070557e-02, -5.3710938e-02, -4.1320801e-02],
...
        [-4.0381775e+00, -4.0255127e+00, -4.0128174e+00, ...,
         -1.5777588e-02, -3.1127930e-03,  9.5825195e-03],
        [-4.0385742e+00, -4.0258789e+00, -4.0131836e+00, ...,
         -1.5411377e-02, -2.7160645e-03,  9.9487305e-03],
        [-4.0389404e+00, -4.0262451e+00, -4.0135803e+00, ...,
         -1.5045166e-02, -2.3498535e-03,  1.0314941e-02]],

       [[-3.9858093e+00, -3.9734802e+00, -3.9611206e+00, ...,
         -6.6741943e-02, -5.4412842e-02, -4.2053223e-02],
        [-3.9861755e+00, -3.9738159e+00, -3.9614563e+00, ...,
         -6.6406250e-02, -5.4046631e-02, -4.1687012e-02],
        [-3.9865417e+00, -3.9741821e+00, -3.9618225e+00, ...,
         -6.6070557e-02, -5.3710938e-02, -4.1320801e-02],
        ...,
        [-4.0381775e+00, -4.0255127e+00, -4.0128174e+00, ...,
         -1.5777588e-02, -3.1127930e-03,  9.5825195e-03],
        [-4.0385742e+00, -4.0258789e+00, -4.0131836e+00, ...,
         -1.5411377e-02, -2.7160645e-03,  9.9487305e-03],
        [-4.0389404e+00, -4.0262451e+00, -4.0135803e+00, ...,
         -1.5045166e-02, -2.3498535e-03,  1.0314941e-02]]], dtype=float32)
Coordinates:
    XLONG     (south_north, west_east) float32 186kB -3.986 -3.973 ... 0.01031
    XLAT      (south_north, west_east) float32 186kB 54.24 54.24 ... 55.28 55.28
  * Time      (Time) datetime64[ns] 192B 2024-05-21 ... 2024-05-21T23:00:00
    datetime  (Time) datetime64[ns] 192B 2024-05-21 ... 2024-05-21T23:00:00
Dimensions without coordinates: south_north, west_east
Attributes:
    FieldType:    104
    MemoryOrder:  XY 
    description:  LONGITUDE, WEST IS NEGATIVE
    units:        degree_east
    stagger:      
    coordinates:  XLONG XLAT
    projection:   LambertConformal(stand_lon=-2.0, moad_cen_lat=57.0, truelat...